### PR TITLE
skip `raise` basic blocks in `verifyFanoutBackward`

### LINKED ISSUE #1023: `fanout_raise` in `refpruning` algorithm doesn't work on this subtle case
Continue from https://github.com/numba/numba/issues/9213
Tested with 0.39.1. I feel latest version still has this issue.

I found a subtle case for refcounting, which our refpruning algorithm under `ffi` can't prune it.
However, it should be in theory, since I think any path reaching `raise` block can be ignored?

image

Just want to make sure that you also think this's prunable. 
I do have a real case having this kind of pattern and have a workaround for this. Removing these refcounting calls can boost 5%~10% performance for my benchmark.

cc @sklam

### PR DESCRIPTION
**Pull request by dlee992**
I provide a potential fix for #1023 .
But I can't prove it in theory, maybe you can give a counterexample against this PR.
Will provide a test case for this.

fix #1023

### CODE REVIEW

File: llvmlite/tests/test_refprune.py [hunk 1]
```diff
@@ -521,6 +521,37 @@ def test_fanout_raise_5(self):
         mod, stats = self.check(self.fanout_raise_5)
         self.assertEqual(stats.fanout_raise, 2)
 
+    # test case 6 is from https://github.com/numba/llvmlite/issues/1023
```
Comment (gmarkall) [inline llvmlite/tests/test_refprune.py:524]: There's no IR or even Python source anywhere in https://github.com/numba/llvmlite/issues/1023 - perhaps this link should be pointing somewhere else?

**Review by gmarkall — COMMENTED** [review]
I rebuilt llvmlite locally with this patch, and the Numba test suite passed for me.

File: llvmlite/tests/test_refprune.py (hunk 1 above)
Comment (dlee992) [reply llvmlite/tests/test_refprune.py:524]: Emm, I mean this test case is from the graph in #1023.

**Comment by dlee992** [comment]
> I rebuilt llvmlite locally with this patch, and the Numba test suite passed for me.

Thanks, graham! This can be marked as `ready for review` I think.

File: ffi/custom_passes.cpp [hunk 2]
```diff
@@ -864,6 +887,10 @@ struct RefPrunePass : public FunctionPass {
             while (workstack.size() > 0) {
                 // Get a basic block
                 BasicBlock *cur_node = workstack.pop_back_val();
+                // If cur_node is a raising block, then skip it
+                if (isRaising(cur_node)) {
+                    continue;
+                }
```
Comment (sklam) [inline ffi/custom_passes.cpp:893]: This needs to respect the `prune_raise_exit` flag from `findFanout()`.

File: llvmlite/tests/test_refprune.py (hunk 1 above)
Comment (sklam) [reply llvmlite/tests/test_refprune.py:524]: Is this pattern generated by the reproducer in https://github.com/numba/numba/issues/9213#issue-1911942120? I don't see this patch to have any effect to that.

**Review by sklam — COMMENTED** [review]
I will need a second pass to think through this change.

**Comment by dlee992** [comment]
RE: https://github.com/numba/llvmlite/pull/1025#discussion_r1458090545
No, this pattern comes from #1023. I didn't share the real code for this, since it's too big, and don't how to get an example from it.
But I think this patch is kind of general.

File: ffi/custom_passes.cpp (hunk 2 above)
Comment (dlee992) [reply ffi/custom_passes.cpp]: addressed this suggestion. Now check `prune_raise_exit` as well.

**Comment by dlee992** [comment]
Has a format issue. Will install pre-commit to solve it

**Comment by dlee992** [comment]
fixed the C++ format issue.

**Comment by dlee992** [comment]
Gentle cc @sklam , ready for review again.

**Comment by dlee992** [comment]
And how about put this small fix into next release?

**Comment by dlee992** [comment]
Can this be marked with next release milestone?

**Review by sklam — APPROVED** [review]
Passed the numba testsuite.

**Pull request merged**